### PR TITLE
Add docker security headers via labels

### LINKED ISSUE #1903: SSLRedirect = true / Traefik Label for Docker Swarm
### Feature

We basically looking to force certain containers to use SSL. Possible syntaxes are:

1. `traefik.frontend.headers=SSLRedirect:true`
2. `traefik.headers=SSLRedirect:true`
3. `traefik.frontend.headers.SSLRedirect=true`
4. `traefik.headers.SSLRedirect=true`

### Context

I can't find an issue regarding this so I prefer to open an issue to follow this feature.

The conversation below happened in Slack with @dtomcej.
I want copy it here because Slack history gets hiden over time :-/

### Slack conversation

<details>

**dtomcej [10:45 AM]**
I just finished a PR to include a bunch of security features
Such as security headers, etc
but it also enabled SSL redirection per-frontend

**pascalandy [10:45 AM]**
Ok

**dtomcej [10:46 AM]**
and I don’t think it strips paths

**pascalandy [10:46 AM]**
As you can see here https://github.com/pascalandy/docker-stack-this/tree/edge-acme/traefik-manager
I want to build a solid Docker Stack for docker swarm mode

This is solid but there is no ACME

**dtomcej [10:47 AM]**
ya,
```toml
[entryPoints.http.redirect]
  entryPoint = "https"
```
uses a regex based redirection

Can you try with this container: `containous/traefik:f275e4ad3c0a660aa8170d0f5f709df5a9f7ebf4`

**pascalandy [10:48 AM]**
instead of `containous/traefik:1.3.0` right ?

**dtomcej [10:48 AM]**
correct

**pascalandy [10:48 AM]**
ok

**dtomcej [10:48 AM]**
that is an intermediate build off of master
it should allow you to add some other flags to your Toml

```toml
[frontends]
  [frontends.frontend1]
  backend = "backend1"
    [frontends.frontend1.routes.test_1]
    rule = "PathPrefixStrip:/cheddar"
  [frontends.frontend2]
  backend = "backend2"
    [frontends.frontend2.headers]
    SSLRedirect = true
    [frontends.frontend2.routes.test_1]
    rule = "PathPrefixStrip:/stilton"
```
 
that commit is here: https://github.com/containous/traefik/commit/f275e4ad3c0a660aa8170d0f5f709df5a9f7ebf4

with all the features that were added

**pascalandy [10:51 AM]**
https://github.com/pascalandy/docker-stack-this/commit/c1c3ab9eb823b2cbe4bb72905fd92071dccc4087

**dtomcej [10:51 AM]**
but the
```toml
[frontends.frontend2.headers]
    SSLRedirect = true
```
is the one you should test

yes

but there should also be another label you can add to your containers

**pascalandy [10:51 AM]**
OLD
```yml
  a-cad:
    image: abiosoft/caddy
    networks:
      - ntw_front
    deploy:
      replicas: 2
      placement:
        constraints: [node.role==manager]
      restart_policy:
        condition: on-failure
      resources:
        limits:
          cpus: ‘0.10’
          memory: 15M
        reservations:
          memory: 9M
      labels:
        - "traefik.backend=a-cad"
        - "traefik.frontend.rule=PathPrefixStrip:/a-cad"
        - "traefik.port=2015"
        - "traefik.docker.network=ntw_front"
        - "traefik.backend.loadbalancer.method=wrr"
```

NEW
```yml
  a-cad:
    image: abiosoft/caddy
    networks:
      - ntw_front
    deploy:
      replicas: 2
      placement:
        constraints: [node.role==manager]
      restart_policy:
        condition: on-failure
      resources:
        limits:
          cpus: ‘0.10’
          memory: 15M
        reservations:
          memory: 9M
      labels:
        - "traefik.backend=a-cad"
        - "traefik.frontend.rule=PathPrefixStrip:/a-cad"
        - "traefik.port=2015"
        - "traefik.docker.network=ntw_front"
        - "traefik.backend.loadbalancer.method=wrr"
        - "traefik.SSLRedirect=true"
```

**dtomcej [10:52 AM]**
See if `-- "traefik.frontend.headers=SSLRedirect:true"` works

I think thats how the toml parses

**pascalandy [10:56 AM]**
like this https://github.com/pascalandy/docker-stack-this/commit/af03cc697cf45ed1f445a91a4ef55bd9a6493cc7

**dtomcej [10:56 AM]**
ya
 
give that a try
 
see if it works

**pascalandy [10:57 AM]**
deploying …

nope!

**dtomcej [10:57 AM]**
also you should remove the entrypoint redirect

no didnt parse properly?

**pascalandy [10:57 AM]**
the webapp are not showing up at all

**dtomcej [10:57 AM]**
in traefik?
or just via curl

**pascalandy [10:58 AM]**
in docker
 
```
ID                  NAME                MODE                REPLICAS            IMAGE                                                         PORTS
3n55qbzzoglo        10000016-caddy      replicated          1/1                 abiosoft/caddy:0.10.3
ag9zzjj4vj6b        proxy_traefik       replicated          1/1                 containous/traefik:f275e4ad3c0a660aa8170d0f5f709df5a9f7ebf4   *:80->80/tcp,*:443->443/tcp,*:8080->8080/tcp
atnk4ulmguvb        10000017-ghost      replicated          1/1                 ghost:alpine
h7fsy0gkbhfu        10000016-ghost      replicated          1/1                 ghost:alpine
v12mgjo9nsd1        tool-resilio        global              1/1                 devmtl/resilio-fire:2.4.4-alpine-0                            *:33333->33333/tcp
```

**dtomcej [10:59 AM]**
what does the traefik debug log say?

the json blob shows the parsed configuration

**pascalandy [10:59 AM]**
First, I show remove this
```toml
[entryPoints.http.redirect]
  entryPoint = "https"
```
from `traefik.toml`

**dtomcej [10:59 AM]**
yes

**pascalandy [11:00 AM]**
this will be a cleaner solution :slightly_smiling_face:

**dtomcej [11:00 AM]**
it will also allow SOME apps to force ssl

but not otheres

**pascalandy [11:01 AM]**
exactly!

specially when developping

**dtomcej [11:01 AM]**
:smile:

**pascalandy [11:02 AM]**
where is the other version of `-- "traefik.frontend.headers=SSLRedirect:true"` ?

**dtomcej [11:02 AM]**
?

**pascalandy [11:03 AM]**
This is not working `-- "traefik.frontend.headers=SSLRedirect:true"`

you proposed another way

**dtomcej [11:03 AM]**
you could try `-- "traefik.headers=SSLRedirect:true"`

Im not sure how the docker labels are parsed (I think the frontend scope is removed)

**pascalandy [11:07 AM]**
wait for it ..

no redirect for https://cloudsite.com/a-gix
but at least cloudsite.com/a-gix works as well

**dtomcej [11:10 AM]**
was `/a-gix` the url you were testing with?

`/a-cad` was your one from before

**pascalandy [11:11 AM]**
I run 4 containers
https://github.com/pascalandy/docker-stack-this/tree/edge-acme/traefik-manager#see-our-services

a-gix is nginx

a-cad is caddy
 
etc

**dtomcej [11:12 AM]**
ya, Im curious to see what the traefik debug logs are for the container that you have the `SSLRedirect:true` for

Its possible that flag might not be parsing correctly

(I don’t use docker personally, so I’m not intimately familiar with the label parsing)

pascalandy [11:14 AM] 
there you go! LOGS https://gist.github.com/pascalandy/83022e9d6e44d799626f1f1236d45a7b

```
3n55qbzzoglo        10000016-caddy      replicated          1/1                 abiosoft/caddy:0.10.3
h7fsy0gkbhfu        10000016-ghost      replicated          1/1                 ghost:alpine
atnk4ulmguvb        10000017-ghost      replicated          1/1                 ghost:alpine
i7nyfcrgh2al        proxy_traefik       replicated          1/1                 containous/traefik:f275e4ad3c0a660aa8170d0f5f709df5a9f7ebf4   *:80->80/tcp,*:443->443/tcp,*:8080->8080/tcp
v12mgjo9nsd1        tool-resilio        global              1/1                 devmtl/resilio-fire:2.4.4-alpine-0                            *:33333->33333/tcp
b1o4haahugqb        webapp_a-cad        replicated          2/2                 abiosoft/caddy:latest
kezzqbedcevw        webapp_a-der        replicated          2/2                 jwilder/whoami:latest
3evd4zwlq2d7        webapp_a-gix        replicated          2/2                 nginx:alpine
busanj2rblhf        webapp_a-who        replicated          2/2                 emilevauge/whoami:latest
181s2cfiv0t3        webapp_b-cad        replicated          2/2                 abiosoft/caddy:latest
13p8fbgv3oho        webapp_b-der        replicated          2/2                 jwilder/whoami:latest
ds6aume5qa07        webapp_b-gix        replicated          2/2                 nginx:alpine
fn2qafnnwdn6        webapp_b-who        replicated          2/2                 emilevauge/whoami:latest

```

As you can see here, the parsing is clean
https://github.com/pascalandy/docker-stack-this/blob/edge-acme/traefik-manager/webapp.yml#L26 (edited)

**dtomcej [11:17 AM]**
```json
"frontend-PathPrefixStrip-b-who":{"entryPoints":["http","https"],"backend":"backend-b-who","routes":{"route-frontend-PathPrefixStrip-b-who":{"rule":"PathPrefixStrip:/b-who"}},"passHostHeader":true,"priority":0,"basicAuth":[],"headers":{}}}}"
```

**pascalandy [11:17 AM]**
or …
` - "traefik.headers.sslredirect=true" `
would make sense

**dtomcej [11:17 AM]**
That log shows the headers struct is not being parsed

**pascalandy [11:17 AM]**
ok

**dtomcej [11:19 AM]**
hmmm

it might be `traefik.frontend.headers.SSLRedirect = true`

**pascalandy [11:20 AM]**
looks better

let me try
```yml
labels:
- "traefik.backend=a-cad"
- "traefik.frontend.rule=PathPrefixStrip:/a-cad"
- "traefik.frontend.headers.SSLRedirect=true"
- "traefik.port=2015"
- "traefik.docker.network=ntw_front"
- "traefik.backend.loadbalancer.method=wrr"
```

**pascalandy [11:28 AM]**
To give you the full picture of this Docker Stack. 8 containers with 2 replicas each:

```
http://cloudsite.com/a-gix
http://cloudsite.com/a-cad
http://cloudsite.com/a-who
http://cloudsite.com/a-der

http://cloudsite.com/b-gix
http://cloudsite.com/b-cad
http://cloudsite.com/b-who
http://cloudsite.com/b-der

https://cloudsite.com/a-gix
https://cloudsite.com/a-cad
https://cloudsite.com/a-who
https://cloudsite.com/a-der

https://cloudsite.com/b-gix
https://cloudsite.com/b-cad
https://cloudsite.com/b-who
https://cloudsite.com/b-der
```

I get the same result

**dtomcej [11:29 AM]**
do the logs show that header field being parsed?

**pascalandy [11:31 AM]**
Still not :disappointed: https://gist.github.com/pascalandy/ed2da0fca1451ed59d0bc8ae655e02e8

Just to make sure, here is how my webapps are looking like:
https://github.com/pascalandy/docker-stack-this/commit/07a08e17883666be97390fd224ff1c184dd6a582

I use `caServer = "https://acme-staging.api.letsencrypt.org/directory"` at the moment
as I already have certificates for cloudsite.com

I don’t think it could cause issue

**dtomcej [11:36 AM]**
naw, that shouldnt be an issue

hmmm

**pascalandy [11:37 AM]**
EDIT2 It does not redirect automatically but at least both http and https are working in parralele.

**dtomcej [11:39 AM]**
hmmmm I am still confused as to why its not parsing

**pascalandy [11:40 AM]**
We could try a docker service
 
instead of a docker compose stack (edited)

**dtomcej [11:40 AM]**
but the other labels are parsing

**pascalandy [11:40 AM]**
I know

To conclude we tried these combinaisons:
```
"traefik.frontend.headers=SSLRedirect:true"
"traefik.headers=SSLRedirect:true"
"traefik.frontend.headers.SSLRedirect=true"
```

**dtomcej [11:42 AM]**
ya

the only last one to try is

`traefik.headers.SSLRedirect = true`

other than that, im super confused, and will have to go back to my dev machine to run some more tests regarding parsing

**pascalandy [11:46 AM]**
If I compare to `"traefik.frontend.passHostHeader=true"`

This would make sense: `"traefik.frontend.headersSSLRedirect=true"`

**dtomcej [11:47 AM]**
It should right?

**pascalandy [11:48 AM]**
or..

`"traefik.frontend.SSLRedirectheaders=true"`

**dtomcej [11:49 AM]**
But headers is a struct INSIDE frontend

https://github.com/containous/traefik/blob/master/types/types.go#L119
 
and SSLRedirect is a key inside that header struct

**pascalandy [11:51 AM]**
I see. So this
`3. "traefik.frontend.headers.SSLRedirect=true"`
or
`4. "traefik.headers.SSLRedirect=true"`

```
# Example
# "traefik.frontend.passHostHeader=true"

# Testing
1. "traefik.frontend.headers=SSLRedirect:true"
2. "traefik.headers=SSLRedirect:true"
3. "traefik.frontend.headers.SSLRedirect=true"
4. "traefik.headers.SSLRedirect=true"

# Idea
# "traefik.frontend.headersSSLRedirect=true"
# "traefik.frontend.SSLRedirectheaders=true"
```

Let me try 4.

**dtomcej [11:52 AM]**
kk

**pascalandy [11:55 AM]**
webapps are loading …

same behavior

headers is empty

**dtomcej [11:56 AM]**
:disappointed:

**pascalandy [11:56 AM]**
both http and https are working, 
without auto-redirect to https

----- June 16th -----
**pascalandy [11:16 PM]**
Hello Daniel. I open an issue https://github.com/containous/traefik/issues/1766
It’s really funny that Portainer does not work anymore. Cheers!

</details>

### PR DESCRIPTION
**Pull request by dtomcej**
This PR creates the ability to configure the security headers for frontends via docker labels.
This should provide comprehensive functionality such as per-frontend SSL redirection.

Based on some of the work done in https://github.com/containous/traefik/pull/2146

Fixes #1903

### CODE REVIEW

File: provider/docker/docker.go [hunk 1]
```diff
@@ -259,44 +259,80 @@ func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *s
...
+		"getServiceFrontendRule":            p.getServiceFrontendRule,
+		"getServicePassHostHeader":          p.getServicePassHostHeader,
+		"getServicePriority":                p.getServicePriority,
+		"getServiceBackend":                 p.getServiceBackend,
+		"getWhitelistSourceRange":           p.getWhitelistSourceRange,
+		"hasRequestHeaders":                 p.hasRequestHeaders,
+		"getRequestHeaders":                 p.getRequestHeaders,
+		"hasResponseHeaders":                p.hasResponseHeaders,
+		"getResponseHeaders":                p.getResponseHeaders,
+		"hasAllowedHostsHeaders":            p.hasAllowedHostsHeaders,
```
Comment (ldez) [inline provider/docker/docker.go]: You can simplify your code like that:
```
"hasAllowedHostsHeaders":            p.hasLabel(types.LabelFrontendAllowedHosts),
```

```go
func (p Provider) hasLabel(label string) func(container dockerData) bool {
	return func(container dockerData) bool {
		label, err := getLabel(container, label)
		return err == nil && len(label) > 0
	}
}
```

Comment (dtomcej) [reply provider/docker/docker.go]: Good Call!

Was just getting code roughed in. I knew that there is a lot of duplication, but I wanted to see it as a whole first, before condensing :)

**Comment by dtomcej** [comment]
In case it wasn't clear, I will be integrating k8s annotations in this PR as well, since the functions will be the same.

**Comment by ldez** [comment]
So that your PR is integrated in the 1.5, it is imperative that it is ready by Tuesday 21 November at the latest. :rocket:

**Comment by ldez** [comment]
@dtomcej So that your PR is integrated in the 1.5, it is imperative that it is ready by Tuesday 21 November at the latest. :rocket:

**Comment by dtomcej** [comment]
Thank you for the reminder! Will have it ready!

File: types/common_label.go [hunk 2]
```diff
@@ -16,6 +16,24 @@ const (
 	LabelFrontendEntryPoints                     = LabelPrefix + "frontend.entryPoints"
 	LabelFrontendRequestHeader                   = LabelPrefix + "frontend.headers.customrequestheaders"
 	LabelFrontendResponseHeader                  = LabelPrefix + "frontend.headers.customresoponseheaders"
```
Comment (else) [inline types/common_label.go]: I just noticed this typo: `customresoponseheaders`

Comment (ldez) [reply types/common_label.go]: It's already fixed in `master`

File: provider/docker/docker.go [hunk 3]
```diff
@@ -260,46 +260,81 @@ func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *s
...
-		"getServicePort":              p.getServicePort,
-		"getServiceWeight":            p.getServiceWeight,
-		"getServiceProtocol":          p.getServiceProtocol,
-		"getServiceEntryPoints":       p.getServiceEntryPoints,
-		"getServiceBasicAuth":         p.getServiceBasicAuth,
-		"getServiceFrontendRule":      p.getServiceFrontendRule,
-		"getServicePassHostHeader":    p.getServicePassHostHeader,
-		"getServicePriority":          p.getServicePriority,
-		"getServiceBackend":           p.getServiceBackend,
-		"getServiceRedirect":          p.getServiceRedirect,
```
[the remark this answers is not in the document]
Comment (dtomcej) [reply provider/docker/docker.go]: Added!

File: provider/docker/docker.go [hunk 4]
```diff
@@ -843,6 +876,125 @@ func parseCustomHeaders(container dockerData, containerType string) map[string]s
...
+	if len(value) == 0 {
+		log.Errorf("Could not load %v headers", containerType)
+	}
+	return value
+}
+
+func getBoolHeader(container dockerData, containerType string) bool {
+	label, err := getLabel(container, containerType)
+	return err == nil && len(label) > 0 && strings.EqualFold(strings.TrimSpace(label), "true")
+
```
Comment (ldez) [inline provider/docker/docker.go]: Could your remove this empty line?

Comment (dtomcej) [reply provider/docker/docker.go]: Yup!

**Comment by robholland** [comment]
It doesn't look like this adds support for kubernetes?

**Comment by ldez** [comment]
@robholland this PR only concern Docker.

**Comment by robholland** [comment]
https://github.com/containous/traefik/pull/2334#issuecomment-342579446 says kubernetes would be included and https://github.com/containous/traefik/pull/2146 was closed in favour of this :(

**Comment by dtomcej** [comment]
@robholland You are correct. My intention was to have both in this PR, but it was decided to keep the two separate due to testing. Will have a new PR for k8s today. Again, apologies for the delay.

**Comment by dtomcej** [comment]
@robholland Discussed with the team, and the new PR will make it into 1.5, so will make the next release, so this delay will not affect the mainline release :)

**Comment by robholland** [comment]
Great, thanks! :)

File: docs/configuration/backends/docker.md [hunk 5]
```diff
@@ -169,11 +169,33 @@ Labels can be used on containers to override default behaviour.
...
+| `traefik.frontend.headers.allowedHosts=EXPR`              | Provides a list of allowed hosts that requests will be processed. Format: `Host1,Host2`                                                                                                                                                                                                                                                                                                                                                     |
+|`traefik.frontend.headers.customrequestheaders=EXPR `     | Provides the container with custom request headers that will be appended to each request forwarded to the container. Format:  `HEADER:value,HEADER2:value2`                                                                                                                                                                                                            |
```
Comment (vito-c) [inline docs/configuration/backends/docker.md]: are these meant to be lowercase or camel case?

`traefik.frontend.headers.customrequestheaders` should be `traefik.frontend.headers.customRequestHeaders`

File: types/common_label.go [hunk 6]
```diff
@@ -16,6 +16,24 @@ const (
 	LabelFrontendEntryPoints                     = LabelPrefix + "frontend.entryPoints"
 	LabelFrontendRequestHeader                   = LabelPrefix + "frontend.headers.customrequestheaders"
```
Comment (vito-c) [inline types/common_label.go]: if camelcase then these should be updated as well

File: docs/configuration/backends/docker.md (hunk 5 above)
Comment (ldez) [reply docs/configuration/backends/docker.md]: Please open an issue instead of commented an already merged PR.

File: types/common_label.go (hunk 6 above)
Comment (ldez) [reply types/common_label.go]: Please open an issue instead of commented an already merged PR.

File: docs/configuration/backends/docker.md (hunk 5 above)
Comment (ldez) [reply docs/configuration/backends/docker.md]: @vito-c or open a PR. thanks

File: types/common_label.go (hunk 6 above)
Comment (ldez) [reply types/common_label.go]: @vito-c or open a PR. thanks

**Pull request merged**